services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateManager.java

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2019 The Alcor Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+        you may not use this file except in compliance with the License.
+        You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+        Unless required by applicable law or agreed to in writing, software
+        distributed under the License is distributed on an "AS IS" BASIS,
+        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+
+package com.futurewei.alcor.dataplane.utils;
+
+import com.futurewei.alcor.dataplane.service.GoalStateService;
+import com.futurewei.alcor.schema.*;
+import com.futurewei.alcor.schema.Port.PortState;
+import com.futurewei.alcor.web.entity.dataplane.InternalPortEntity;
+import com.futurewei.alcor.web.entity.dataplane.InternalSubnetEntity;
+import com.futurewei.alcor.web.entity.dataplane.NeighborInfo;
+import com.futurewei.alcor.web.entity.dataplane.NetworkConfiguration;
+import com.futurewei.alcor.web.entity.port.PortEntity;
+import com.futurewei.alcor.web.entity.vpc.VpcEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+import static com.futurewei.alcor.schema.Port.PortConfiguration.FixedIp;
+
+@Component
+public class GoalStateManager {
+  @Autowired private GoalStateService goalStateService;
+  private static final Logger LOG = LoggerFactory.getLogger(GoalStateManager.class);
+  private Map<String, Set<String>> portsInSameSubnetMap = new HashMap<>();
+  private Map<String, Set<NeighborInfo>> neighborInfoInSameSubenetMap = new HashMap<>();
+
+  public Map<String, Goalstate.GoalState> transformNorthToSouth(
+      // get the data
+      NetworkConfiguration networkConfiguration) throws RuntimeException {
+    InternalPortEntity[] portStatesArr =
```

Comment:
Didn't see the conversion from List to Array is necessary. If you go through transformNorthToSouth method, the only place to use the array is Line 70 to iterate through the array, which can be done with List as well. @haboy52581

Reply:
I don't understand this why list is better than array

Reply:
I didn't suggest we have to use List. The comment refers to that if we need the conversion or not. 

This type of conversion is not free and requires overhead of copying memory from one place to the other. If it is absolutely needed, of course we should keep it as is. Is it the case here?

Reply:
I think we could leave it for now since in the following code there are codes relying on the array which is easier than arraylist as well , and we could not get any good from arraylist other than looks more object oriented

Reply:
Okay. Consider to optimize this when we refactor the code (tracked by https://github.com/futurewei-cloud/alcor/issues/287)